[새기능] 알림 발송 기능, 알림 조회 기능 구현

jyj1289 opened the pull request:
## 🎫 관련 이슈
[//]: # (다음 키워드를 사용하면 해당 PR을 머지할 때 자동으로 이슈를 닫을 수 있습니다.)
[//]: # (keyword: close|closes|closed|resolve|resolves|resolved|fix|fixes|fixed)
[//]: # (예시: close #1)

close #7 

<br>

## 📄 개요
[//]: # (작업 내용을 간단히 요약해서 적습니다.)
[//]: # (예시: 유저 회원가입 기능을 만들었습니다.)

> 알림 발송, 저장, 조회 기능을 만들었습니다.

<br>

## 🔨 작업 내용
[//]: # (작업 내용을 자세하게 적습니다.)
[//]: # (붙임표 "-" 을 사용해서 목록을 만듭니다.)
[//]: # (예시: 유저 회원가입 API를 만들었습니다.)

- 알림 엔티티를 만들었습니다.
- FCM과 연동하여 알림 발송 및 저장 API를 만들었습니다.
- 알림 조회 API를 만들었습니다.
- 위 2개의 API 테스트 코드를 작성했습니다.
- 위 2개의 API를 문서화했습니다.

<br>

## 🏁 확인 사항
[//]: # (PR을 보내기 전 다음 사항을 확인해주세요.)
[//]: # (해당 사항을 모두 이행해야 머지할 수 있습니다.)
[//]: # (- [x] 를 사용해서 완료로 표시할 수 있습니다.)

- [x] 테스트를 완료했나요?
- [x] API 문서를 작성했나요?
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 로그, 주석, import 등을 삭제했나요?

<br>

## 🙋🏻 덧붙일 말
[//]: # (다음 사항이 있다면 적어주세요.)
[//]: # (PR에 대한 추가 설명)
[//]: # (중점적으로 리뷰받고 싶은 부분)
[//]: # (기타 등등)

jyj1289 committed:
feat(#7): 알림 발송, 저장, FCM 토큰 저장 기능 구현
- FCM을 이용하여 알림 발송기능을 구현했어요.
- 알림을 발송할 때 저장도 같이하도록 구현했어요.
- FCM 토큰을 저장할 수 있도록 구현했어요.
- FCM을 위한 의존성을 추가했어요.

jyj1289 committed:
refactor(#7): 불필요한 예외 삭제
- FCMTokenNotFoundException은 사용하지 않아 제거했어요.

jyj1289 committed:
fix(#7): FCM 토큰 예외 구현
- FCM 토큰이 1개 이상 없을 때 발생하는 예외를 구현했어요.

jyj1289 committed:
refactor(#7): 접근 제어지시자 변경
- DTO에서 굳이 protected까지로 접근 제어지시자를 지정할 이유가 없다고 판단되어 삭제했어요.

jyj1289 committed:
test(#7): 알림 발송 API 테스트 코드 작성
- 알림 발송 API 테스트 코드를 작성했어요.

jyj1289 committed:
docs(#7): 알림 발송 API 문서화
- 알림 발송 API를 문서화했어요.

jyj1289 committed:
test(#7): FCM 토큰 저장 API 테스트 코드 작성
- FCM 토큰 저장 API 테스트 코드를 작성했어요.

jyj1289 committed:
docs(#7): FCM 토큰 저장 API 문서화
- FCM 토큰 저장 API를 문서화 했어요.

jyj1289 committed:
feat(#7): FCM 토큰 삭제 기능 구현
- FCM 토큰 삭제 기능을 구현했어요.

jyj1289 committed:
test(#7): FCM 토큰 삭제 API 테스트 코드 작성
- FCM 토큰 삭제 API 테스트 코드를 작성했어요.

jyj1289 committed:
docs(#7): FCM 토큰 삭제 API 문서화
- FCM 토큰 삭제 API를 문서화했어요.

jyj1289 committed:
feat(#7): 유저별 알림 전체 조회 기능 구현
- 유저별 알림 전체 조회 기능을 구현했어요.

jyj1289 committed:
refactor(#7): 불필요한 impport 삭제
- 불필요한 import문을 제거했어요.

jyj1289 committed:
test(#7): 알림 조회 API 테스트 코드 작성
- 알림 조회 API 테스트 코드를 작성했어요.

jyj1289 committed:
docs(#7): 알림 조회 API 문서화
- 알림 조회 API를 문서화했어요.

jyj1289 committed:
docs(#7): Request Header 추가
- 문서내에 Request Header를 추가했어요.

cabbage16 committed:
Merge branch 'develop' into feat/#7

cabbage16 commented on src/main/java/com/bamdoliro/sinabro/application/fcm/token/DeleteFCMTokenUseCase.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Transactional
+    public void execute(DeleteFCMTokenRequest request) {
+        fcmTokenRepository.deleteByToken(request.getToken());
```
P3) 이렇게만 하면 다른 사용자의 FCMToken이 유출되었을 때, 유저 검증 없이 Token 값만으로 삭제할 수 있어서 위험할 거 같아요.

jyj1289 committed:
feat(#7): FCM Token 저장 및 삭제 시 검증 로직 구현
- FCM Token을 저장하거나 삭제할 때 이미 있는지 확인하는 검증 로직을 구현하였습니다.
- FCMTokenAlreadySavedException을 구현했어요.

cabbage16 committed:
유진이가 자꾸 지가 이상하게 만들어놓고 나보고 고쳐달래요....
ㄴ 밤샘이슈...

jyj1289 committed:
Merge remote-tracking branch 'origin/feat/#7' into feat/#7

jyj1289 committed:
docs(#7): 에러 발생시 응답 문서화
- FCM 토큰 저장 및 삭제에서 에러가 발생했을때 응답을 문서화했어요.
